framework/decode/vulkan_replay_consumer_base.cpp

```diff
@@ -7074,6 +7225,103 @@ VkResult VulkanReplayConsumerBase::OverrideCreateFramebuffer(
+
+            VkCommandBufferAllocateInfo commandBufferAllocateInfo;
+            commandBufferAllocateInfo.sType              = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+            commandBufferAllocateInfo.pNext              = nullptr;
+            commandBufferAllocateInfo.commandPool        = commandPool;
+            commandBufferAllocateInfo.level              = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+            commandBufferAllocateInfo.commandBufferCount = 1;
+
+            VkCommandBuffer commandBuffer;
+            device_table->AllocateCommandBuffers(device, &commandBufferAllocateInfo, &commandBuffer);
+
+            VkCommandBufferBeginInfo commandBufferBeginInfo;
+            commandBufferBeginInfo.sType            = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+            commandBufferBeginInfo.pNext            = nullptr;
+            commandBufferBeginInfo.flags            = VK_COMMAND_BUFFER_USAGE_SIMULTANEOUS_USE_BIT;
+            commandBufferBeginInfo.pInheritanceInfo = nullptr;
+
+            device_table->BeginCommandBuffer(commandBuffer, &commandBufferBeginInfo);
+            device_table->EndCommandBuffer(commandBuffer);
+
+            it = fba_resources_.emplace(device, std::make_pair(commandPool, commandBuffer)).first;
+        }
+
+        // Queue submission with VkFrameBoundaryEXT
+
+        VkPipelineStageFlags dstStageMask = VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT;
+
+        VkSubmitInfo submitInfo;
+        submitInfo.sType                = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+        submitInfo.pNext                = nullptr;
+        submitInfo.waitSemaphoreCount   = 1;
+        submitInfo.pWaitSemaphores      = &semaphore;
+        submitInfo.pWaitDstStageMask    = &dstStageMask;
+        submitInfo.commandBufferCount   = 1;
+        submitInfo.pCommandBuffers      = &it->second.second;
+        submitInfo.signalSemaphoreCount = 0;
+        submitInfo.pSignalSemaphores    = nullptr;
+
+        VkFrameBoundaryEXT frameBoundaryExt;
+        frameBoundaryExt.sType       = VK_STRUCTURE_TYPE_FRAME_BOUNDARY_EXT;
+        frameBoundaryExt.pNext       = nullptr;
+        frameBoundaryExt.flags       = VK_FRAME_BOUNDARY_FRAME_END_BIT_EXT;
+        frameBoundaryExt.frameID     = application_->GetCurrentFrameNumber();
+        frameBoundaryExt.imageCount  = (image == VK_NULL_HANDLE ? 0 : 1);
+        frameBoundaryExt.pImages     = (image == VK_NULL_HANDLE ? nullptr : &image);
+        frameBoundaryExt.bufferCount = 0;
+        frameBoundaryExt.pBuffers    = nullptr;
+        frameBoundaryExt.tagName     = frameBoundaryExt.frameID;
+        frameBoundaryExt.tagSize     = 0;
+        frameBoundaryExt.pTag        = nullptr;
+
+        submitInfo.pNext = &frameBoundaryExt;
+
+        VkQueue queue;
+        device_table->GetDeviceQueue(device, queueFamily, 0, &queue);
+        device_table->QueueSubmit(queue, 1, &submitInfo, VK_NULL_HANDLE);
```

Comment:
One of the big concerns I have here is something we ran into before with the virtual swapchain code.  We were running into a race condition when code was being submitted into a different Queue than the triggering semaphore completion work.  This resulted in us performing the virtual swapchain update with incorrect content.  Sometimes "future" content so we saw frame stuttering.

In this case, you may be receiving the semaphore from work that executed a queue that is not index 0.  If this is the case, you're adding additional code to queue 0 potentially well after the work was done.  This may result in the application already sending pending work onto that queue and the resultant "end of frame" may actually be after this content which correctly belongs to the next frame.